### PR TITLE
add support for safety checker (NSFW filter)

### PR DESCRIPTION
**Pull request by lstein**
Now you can activate the Hugging Face `diffusers` library safety check for NSFW and other potentially disturbing imagery.

To turn on the safety check, pass `--safety_checker` to `invoke.py` on the command line. For developers, the flag is `safety_checker=True` passed to `ldm.generate.Generate()`. Once the safety checker is turned on, it cannot be turned off unless you reinitialize a new Generate object. This PR does not add a safety checker toggle to the WebGUI.

When the safety checker is active, suspect images will be blurred and a warning icon is added. There is also a warning message printed in the CLI, but it can be a little hard to see because of its positioning in the output stream.

Example blurred image:

There is a slight but noticeable delay when the safety checker runs.

Note that invisible watermarking is *not* currently implemented. The watermark code distributed by the CompViz distribution uses a library that does not seem to be able to retrieve the watermarks it creates, and it does not appear that Hugging Face `diffusers` or other SD distributions are doing any watermarking.

### CODE REVIEW

**Comment by tildebyte** [comment]
@lstein; What is the 'diffusers' dependency being used for? I don't see it imported anywhere.

File: scripts/preload_models.py [hunk 1]
```diff
@@ -17,41 +17,39 @@
 
 transformers.logging.set_verbosity_error()
 
+#---------------------------------------------
 # this will preload the Bert tokenizer fles
```
Comment (tildebyte) [inline scripts/preload_models.py:21]: Why so many changes unrelated to the safety checker in here?

File: ldm/invoke/generator/inpaint.py [hunk 2]
```diff
@@ -142,7 +142,7 @@ def sample_to_image(self, samples)->Image:
         # Blur the mask out (into init image) by specified amount
         if mask_blur_radius > 0:
             nm = np.asarray(pil_init_mask, dtype=np.uint8)
-            nmd = cv.dilate(nm, kernel=np.ones((3,3), dtype=np.uint8), iterations=int(mask_blur_radius / 2))
+            nmd = cv.erode(nm, kernel=np.ones((3,3), dtype=np.uint8), iterations=int(mask_blur_radius / 2))
```
Comment (lstein) [inline ldm/invoke/generator/inpaint.py:145]: This line got in here by mistake, but will be removed during merge.

File: scripts/preload_models.py [hunk 3]
```diff
@@ -17,41 +17,39 @@
...
-print('Loading bert tokenizer (ignore deprecation errors)...', end='')
-with warnings.catch_warnings():
-    warnings.filterwarnings('ignore', category=DeprecationWarning)
-    tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased')
-print('...success')
-sys.stdout.flush()
+def download_bert():
```
Comment (lstein) [inline scripts/preload_models.py:22]: While adding the download function for the safety checker model, I felt it was important to reorganize `preload_models.py` to make it more modular (or at least understandable).

File: ldm/invoke/generator/base.py [hunk 4]
```diff
@@ -169,6 +180,39 @@ def slerp(self, t, v0, v1, DOT_THRESHOLD=0.9995):
 
         return v2
 
+    def safety_check(self,image:Image.Image):
+        '''
+        If the CompViz safety checker flags an NSFW image, we
+        blur it out.
+        '''
+        import diffusers
```
Comment (lstein) [inline ldm/invoke/generator/base.py:188]: Here is where `diffusers` is being imported.

**Comment by lstein** [comment]
> @lstein; What is the 'diffusers' dependency being used for? I don't see it imported anywhere.

It is imported in the `safety_checker()` method of `generator/base.py`. It takes a while to load, and will not be used by everyone, so I thought it better to defer loading until needed.

File: scripts/preload_models.py (hunk 3 above)
Comment (tildebyte) [reply scripts/preload_models.py:22]: OK, fair.

Usually one would at least put this in its own commit.

**Review by tildebyte — APPROVED** [review]

**Comment by lstein** [comment]
Merged from command line.

**Comment by ausbitbank** [comment]
Is it possible to have the safety checker return a nsfw flag via the api instead of replacing the image? For my use case as a discord bot via api it would be ideal to still get the nsfw render and simply hide the imag behind a spoiler tag that requires clicking to see it (or in a seperate nsfw channel etc).

The idea of throwing away "expensive" renders based on results from a inaccurate nsfw check (in the SD beta test version it regularly flagged SFW and ignored NSFW) feels wrong. Sorry to bring it up after you've already merged, I only just saw the thread... My $0.02

**Pull request closed without merging**